File: node-app/src/lib/ai/philosophy.ts
// src/lib/ai/philosophy.ts
import { getPipeline } from "./models";
import { getTokenizer } from "./tokenizer";

/**
 * Generate a short philosophical or thematic interpretation of an article.
 *
 * Uses a summarization model to create a compressed representation of the
 * article’s emotional or conceptual tone, then reframes it as a
 * philosophical insight.
 *
 * This is an approximation until a dedicated model, rhetorical classifier,
 * or LLM-powered interpretive layer is added.
 *
 * ## Processing Logic
 * 1. Tokenize input using GPT-2 tokenizer.
 * 2. Truncate to 512 tokens.
 * 3. Decode tokens and pass to BART summarizer.
 * 4. Prefix output with `"Philosophical framing: "`.
 *
 * ## Returns
 * A short reflective or thematic interpretation, e.g.:
 * ```ts
 * The narrative reflects a broader tension
 * between technological progress and collective responsibility...
 * ```
 *
 * @param article - Partial article containing optional raw content.
 * @returns A short interpretive/philosophical summary string.
 *
 * @example
 * const insight = await generatePhilosophicalInsight(article);
 * console.log(insight);
 */
export async function generatePhilosophicalInsight(article: {
	content?: string;
}): Promise<string> {
	const { content } = article ?? "";
	if (!content || content.trim().length === 0) {
		console.warn("Tokenizer skipped: empty input");
		return "";
	}

	const TOKEN_LIMIT = 512;

	const tokenizer = await getTokenizer();
	const tokens = tokenizer.encode(content);
	if (tokens.length === 0) {
		console.warn("Tokenizer produced empty token array. Using fallback.");
		return "";
	}

	const truncatedTokens = tokens.slice(0, TOKEN_LIMIT);
	if (!truncatedTokens.length) {
		console.warn("Tokenizer truncated to zero tokens, skipping summarization.");
		return "";
	}

	const inputText = tokenizer.decode(truncatedTokens);

	const summarizer = await getPipeline("summarization", "distilbart-cnn");
	const result = await summarizer(inputText, { max_length: 80 });

	console.log(
		`Generating philosophical insight for ${tokens.length} tokens (truncated to ${truncatedTokens.length})`,
	);

	return result[0].summary_text;
}

File: python-ai/src/ai/philosophy.py
"""
philosophy.py

Generate a short philosophical or thematic interpretation of an article.

Uses the shared get_pipeline() helper for summarization.
"""

from typing import Optional, Dict, Any, List
from .models import get_pipeline
from .tokenizer import get_tokenizer
import logging

logger = logging.getLogger(__name__)
logger.setLevel(logging.INFO)

MAX_INPUT_TOKENS = 512
MAX_SUMMARY_TOKENS = 80


def generate_philosophical_insight(content: Optional[str]) -> Dict[str, Any]:
    """
    Generate a short philosophical or thematic interpretation of the article content.

    Returns:
        {
            "status": "ok" | "error",
            "data": { "insight": str },
            "errors": [...],
            "meta": {...}
        }
    """
    errors: List[str] = []
    meta: Dict[str, Any] = {}
    insight: str = ""

    if not content or not content.strip():
        return {
            "status": "error",
            "data": {"insight": ""},
            "errors": ["Empty input provided."],
        }

    # Load tokenizer
    try:
        tokenizer = get_tokenizer()
    except Exception as e:
        msg = f"Tokenizer load failed: {e}"
        logger.error(msg)
        return {
            "status": "error",
            "data": {"insight": ""},
            "errors": [msg],
        }

    # Load summarization pipeline
    try:
        summarizer = get_pipeline("summarization")
        if summarizer is None:
            raise RuntimeError("Summarization pipeline returned None")
    except Exception as e:
        msg = f"Summarization pipeline load failed: {e}"
        logger.error(msg)
        return {
            "status": "error",
            "data": {"insight": ""},
            "errors": [msg],
        }

    # Tokenize + truncate
    try:
        tokens = tokenizer.encode(content)
        meta["original_token_count"] = len(tokens)

        truncated_tokens = tokens[:MAX_INPUT_TOKENS]
        meta["truncated_token_count"] = len(truncated_tokens)

        input_text = tokenizer.decode(truncated_tokens, skip_special_tokens=True)

        if not input_text:
            raise RuntimeError("Tokenizer decode produced empty text")

    except Exception as e:
        msg = f"Tokenization failed: {e}"
        return {
            "status": "error",
            "data": {"insight": ""},
            "errors": [msg],
            "meta": meta,
        }

    # Generate "philosophical insight" summary
    try:
        result = summarizer(
            input_text,
            max_length=MAX_SUMMARY_TOKENS,
            min_length=20,
            do_sample=False
        )
        summary_text = result[0]["summary_text"]
        insight = f"Philosophical framing: {summary_text}"

        meta["insight_length"] = len(insight)

    except Exception as e:
        msg = f"Summarization failed: {e}"
        errors.append(msg)

    status = "ok" if not errors else "error"

    return {
        "status": status,
        "data": {"insight": insight},
        "errors": errors if errors else None,
        "meta": meta if meta else None,
    }
